### ISSUE TITLE
Input collection is overwritten with nil values

### ISSUE THREAD

**Issue by korthout** [camunda-cloud/zeebe#7330]
**Describe the bug**

When both the input collection and output collection of a multi-instance element refer to the same variable, then the input collection's values get overwritten with nil values.

This happens because of the nil initialization of the output collection.

**To Reproduce**

1. Deploy a process with a multi-instance element with the following properties:
- Input Collection: `= list`
- Input Element: `value`
- Output Collection: `list`
- Output Element: `= value`
2. Create an instance of it with the following variables: `{"list":[1,2,3]}`
3. Note the `list` and `value` variables in Operate
- At root scope:
![Screen Shot 2021-06-21 at 11 58 53](https://user-images.githubusercontent.com/3511026/122744230-1a168200-d288-11eb-82de-ad1ccd5f2c69.png)
- At multi-instance element scope:
![Screen Shot 2021-06-21 at 11 59 03](https://user-images.githubusercontent.com/3511026/122744248-1d117280-d288-11eb-9ac4-73dcd2b0cf21.png)

**Expected behavior**

The nil initialization of the output collection should not overwrite values of an already existing variable.

**Environment:**
 - OS: Camunda Cloud
 - Zeebe Version: 1.0.0, 1.0.1

**Comment by saig0** [comment]
## Analyze

To be precise, the input collection variable is not overridden. It is declared in the root scope of the process instance. 

The output collection variable is created in the scope of the multi-instance body. If the name of the output collection variable is the same as for the input collection then it shadows the input collection. When a new inner instance is created then it reads the variable with the name of the input collect. In this case here, it reads the local variable of the multi-instance body (from the output collection) instead of the variable on the root scope.   

In the [documentation](https://docs.camunda.io/docs/1.0/reference/bpmn-processes/multi-instance/multi-instance#collecting-the-output), it is defined that the output variable is created in the scope of the multi-instance body. The variable gets propagated when completing the multi-instance body.

The behavior was (slightly) different in version `0.26` because of a different ordering of the record. It activated the inner instance before setting the output variable. But it never worked for sequential multi-instance. 

To sum it up, I think it is not a valid or recommended use case to use the same variable name.

**Comment by korthout** [comment]
Thanks for the analysis @saig0. 

Let's make sure to update the documentation so that users don't fall in this easy to make mistake.

**Comment by korthout** [comment]
I'll close it, since this is expected behavior. An issue for documentation improvements is created in #288.

**Comment by ajeans** [comment]
@saig0 

> The behavior was (slightly) different in version 0.26 because of a different ordering of the record. It activated the inner instance before setting the output variable. But it never worked for sequential multi-instance.

Are you 100% sure of that? We were using sequential multi instance in `0.26.2` and our integration tests were working fine.

**Comment by ajeans** [comment]
> @saig0
> 
> > The behavior was (slightly) different in version 0.26 because of a different ordering of the record. It activated the inner instance before setting the output variable. But it never worked for sequential multi-instance.
> 
> Are you 100% sure of that? We were using sequential multi instance in `0.26.2` and our integration tests were working fine.

Sorry @saig0 , I confused which part of our workflow showed the problem, this is indeed with a parallel multi instance...

Thanks for the analysis... I am not sure how we will be able to "enrich" an item of collection, we will have a look

**Comment by korthout** [comment]
In-place updating of the items of a collection is not possible with only a multi-instance element. You'll always need a different variable than the input collection to collect the output into. However, you can use this output collection variable and use it to overwrite the original collection variable using an output mapping.

To do this you need to wrap the the multi-instance element with an embedded subprocess. The output mapping of the subprocess can then take the output collection of the multi instance as source and use it to overwrite the original collection. Does that make sense? It would look like this:

<img width="866" alt="Screen Shot 2021-06-23 at 16 04 00" src="https://user-images.githubusercontent.com/3511026/123110718-a2d61f00-d43c-11eb-876c-cede34474955.png">

Note that I also added an input mapping on the subprocess to initialize the output collection locally, so that it doesn't leak (i.e. get's propagated) to the root process scope.

**Comment by B-hamza** [comment]
Thanks for the analysis.
So just to understand, using different names may resolve the problem ? So can we still get the ouput parameters of the task to the outputCollection of the multiInstance scope ? 
<img width="550" alt="Capture d’écran 2021-06-28 à 14 37 37" src="https://user-images.githubusercontent.com/9680921/123638380-6846fa80-d81f-11eb-8cfd-ee1e477bf035.png">
<img width="618" alt="Capture d’écran 2021-06-28 à 14 37 43" src="https://user-images.githubusercontent.com/9680921/123638403-6e3cdb80-d81f-11eb-9fa8-ef37d1d86b57.png">

We still get the output and the outputCollection as null. I may have misunderstood the explanation ?

**Comment by korthout** [comment]
Hi @B-hamza,

It's not entire clear in your screenshot what elements you're placing the input/output mappings on. 

But assuming the multi-instance properties you show in the second screenshot I propose to wrap that multi-instance element in an embedded sub process (like I showed in my screenshot). Then on that sub process you add the following input/output mapping:

<img width="820" alt="Screen Shot 2021-06-28 at 15 20 03" src="https://user-images.githubusercontent.com/3511026/123643092-52880400-d824-11eb-8bab-6c359894bf73.png">

What does this do? 

First, the input mapping will initialize the `outputCollection` variable as a local variable (this makes sure it is not leaked to the root process level), note that this is optional but makes sure your process is not filled with no longer used variables, i.e. it's cleaner.

Second, the output mapping will copy the value of `outputCollection` into the `inputCollection` variable (overwriting it). In the steps after the sub process you'll now have the inputCollection variable available with the updated values.

Please let me know if that helps 🙂

**Comment by B-hamza** [comment]
Sorry I may not be clear in my use case. i'll try with the bpmn, just for a simple task. with different names for inputs and outputs.

We initialized the inputCollection that we use in the multiInstance scope with inputElement. As shown in the screenshot :

<img width="984" alt="Capture d’écran 2021-06-28 à 16 58 48" src="https://user-images.githubusercontent.com/9680921/123658721-46a33e80-d832-11eb-915e-d34b8c68d75d.png">

My parameters for every instance scope are as follows : 

<img width="894" alt="Capture d’écran 2021-06-28 à 17 02 51" src="https://user-images.githubusercontent.com/9680921/123659199-b6b1c480-d832-11eb-98a4-e09dd06a9fca.png">

or even using a constant or inputElement : 
<img width="272" alt="Capture d’écran 2021-06-28 à 17 04 15" src="https://user-images.githubusercontent.com/9680921/123659372-e791f980-d832-11eb-88dc-571735c4317c.png">

Here we got the ouput always to null, what i haven't understood maybe, is the relation between the ouputParameters and the outputCollections. What comes first ? I think the use case should work as the variable scopes describe or maybe I'm missing something 😅.

**Comment by korthout** [comment]
Ok, so input and output mappings on the multi-instance service task will be performed for each of the tasks. So this should work correctly.

However, since you're using parallel multi-instance you'll have to pay attention to how you complete the jobs. Parallel multi-instance activities (like in any concurrent program) can lead to race conditions. Example, the jobs of your service task are completed by a job worker. When it completes these jobs does it pass along all the variables belonging to the job? If you do so, all variables you complete the job with, will overwrite existing variables in your process. This is also the case for the outputCollection. This race condition is described in more detail [here](https://docs.camunda.io/docs/reference/bpmn-processes/multi-instance/multi-instance#variable-mappings). 

So my question to you is: When you complete the jobs for these tasks, do you pass along variables? And if so, please try only passing variables along with the job complete command that have been changed by the processing of the job in the job worker.
